crates/matrix-sdk/src/event_cache/linked_chunk/as_vector.rs

```diff
@@ -0,0 +1,529 @@
+// Copyright 2024 The Matrix.org Foundation C.I.C.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::{
+    collections::VecDeque,
+    ops::ControlFlow,
+    sync::{Arc, RwLock},
+};
+
+use eyeball_im::VectorDiff;
+
+use super::{
+    updates::{ReaderToken, Update, UpdatesInner},
+    ChunkIdentifier,
+};
+
+/// A type alias to represent a chunk's length. This is purely for commodity.
+pub(super) type ChunkLength = usize;
+
+/// A type that transforms a `Vec<Update<Item, Gap>>` —given by
+/// [`Updates`]— into a `Vec<VectorDiff<Item>>` —this
+/// type—. Basically, it helps to consume a [`LinkedChunk<CAP, Item, Gap>`] as
+/// if it was an [`eyeball::ObservableVector<Item>`].
+pub struct AsVector<Item, Gap> {
+    /// Weak reference to [`UpdatesInner`].
+    ///
+    /// Using a weak reference allows [`Updates`] to be dropped
+    /// freely even if a subscriber exists.
+    updates: Arc<RwLock<UpdatesInner<Item, Gap>>>,
+
+    /// The token to read the updates.
+    token: ReaderToken,
+
+    /// Mapper from `Update` to `VectorDiff`.
+    mapper: UpdateToVectorDiff,
+}
+
+impl<Item, Gap> AsVector<Item, Gap> {
+    /// Create a new [`Self`].
+    ///
+    /// `updates_subscriber` is simply `UpdatesSubscriber`.
+    /// `initial_chunk_lengths` must be pairs of all chunk identifiers with the
+    /// associated chunk length. The pairs must be in the exact same order than
+    /// in [`LinkedChunk`].
+    ///
+    /// # Safety
+    ///
+    /// This method is marked as unsafe only to attract the caller's attention
+    /// on the order of pairs. If the order of pairs are incorrect, the entire
+    /// algorithm will not work properly and is very likely to panic.
+    pub(super) unsafe fn new(
+        updates: Arc<RwLock<UpdatesInner<Item, Gap>>>,
+        token: ReaderToken,
+        initial_chunk_lengths: VecDeque<(ChunkIdentifier, ChunkLength)>,
+    ) -> Self {
+        Self { updates, token, mapper: UpdateToVectorDiff { chunks: initial_chunk_lengths } }
+    }
+
+    /// Take the new updates as [`VectorDiff`].
+    ///
+    /// It returns an empty `Vec` if there is no new `VectorDiff` for the
+    /// moment.
+    pub(super) fn take(&mut self) -> Vec<VectorDiff<Item>>
+    where
+        Item: Clone,
+    {
+        let mut updates = self.updates.write().unwrap();
+        let updates = updates.take_with_token(self.token);
```

Comment:
There is no different concept, I don't understand what you're referring to. I've changed the highlighted code a little bit to avoid having 2 variables with the `updates` name though.

Reply:
`self.updates` is rather some kind of `UpdateManager` / `UpdateDispatcher` / `ObservableUpdates` type, but no worries, I'll make a proposal in form of a PR later :)

Reply:
Renamed to `ObservableUpdates` 👍.